scripts/run_all.py

```diff
@@ -4,86 +4,239 @@
 
 import argparse
 import os
+import socket
 import subprocess
 import time
 from glob import glob
-from typing import Any, List
+from types import TracebackType
+from typing import Any, List, Optional, Tuple, Type, cast
 
 
-def run_py_example(path: str, args: List[str] = []) -> None:
+def run_py_example(path: str, viewer_port: int, wait: bool = True, save: Optional[str] = None) -> Any:
+    args = ["--connect", "--addr", f"127.0.0.1:{viewer_port}"]
+    if save is not None:
+        args += ["--save", str(save)]
+
     process = subprocess.Popen(
         ["python3", "main.py", "--num-frames=30", "--steps=200"] + args,
         cwd=path,
     )
-    returncode = process.wait()
-    print(f"process exited with error code {returncode}")
+    if wait:
+        returncode = process.wait()
+        print(f"process exited with error code {returncode}")
+    return process
 
 
-def run_saved_example(path: str, args: List[str] = []) -> None:
+def run_saved_example(path: str, wait: bool = True) -> Any:
     process = subprocess.Popen(
-        ["cargo", "run", "-p", "rerun", "--all-features", "--", "out.rrd"] + args,
+        ["cargo", "run", "-p", "rerun", "--all-features", "--", "out.rrd"],
         cwd=path,
     )
-    returncode = process.wait()
-    print(f"process exited with error code {returncode}")
+    if wait:
+        returncode = process.wait()
+        print(f"process exited with error code {returncode}")
+    return process
 
 
-def collect_examples() -> List[str]:
-    return [os.path.dirname(entry) for entry in glob("examples/python/**/main.py")]
+def get_free_port() -> int:
+    with socket.socket() as s:
+        s.bind(("", 0))
+        return cast(int, s.getsockname()[1])
 
 
-def start_viewer(args: List[str] = []) -> Any:
-    process = subprocess.Popen(
-        ["cargo", "run", "-p", "rerun", "--all-features", "--"] + args,
-        stdout=subprocess.PIPE,
-    )
-    time.sleep(1)  # give it a moment to start
-    return process
+def collect_examples(fast: bool) -> List[str]:
+    if not fast:
+        return [os.path.dirname(entry) for entry in glob("examples/python/**/main.py")]
+    else:
+        # cherry picked
+        return [
+            "examples/python/deep_sdf",
+            "examples/python/dicom",
+            "examples/python/raw_mesh",
+            "examples/python/clock",
+            "examples/python/api_demo",
+            "examples/python/car",
+            "examples/python/colmap",
+            "examples/python/plots",
+            "examples/python/nyud",
+            "examples/python/text_logging",
+        ]
+
+
+class Viewer:
+    port: int
+    should_close: bool
+    web: bool
+    web_viewer_port: int
+    ws_server_port: int
```

Comment:
So many ports! Let's be even more explicit to avoid confusion
```suggestion
    should_close: bool
    web: bool
    sdk_port: int  # where the logging SDK sends the log stream (where the server receives)
    web_viewer_port: int  # the HTTP port where we serve the web viewer
    ws_server_port: int  # the WebSocket port where we serve the log stream
```

Reply:
Yes, I wanted to be sure that it would never run into conflicts. Previously the examples would sometimes start their own viewers because the viewer assigned to them could not start on the default port. The result would be ~10 extra open viewers that weren't receiving any data.